### PR TITLE
Balder/deprecate old api update api

### PR DESCRIPTION
**Pull request by baldersheim**
@bratseth and @vekterli PR

### CODE REVIEW

**Review by bratseth — APPROVED** [review]

**Comment by bratseth** [comment]
LGTM!

**Review by vekterli — COMMENTED** [review]
:+1: just a few comments/questions

File: document/src/main/java/com/yahoo/document/DocumentUpdate.java [hunk 1]
```diff
@@ -248,16 +280,24 @@ public FieldUpdate getFieldUpdate(String fieldName) {
     /**
      * Assigns the field updates of this document update.
      * This document update receives ownership of the list - it can not be subsequently used
-     * by the caller. The list may not be unmodifiable.
+     * by the caller. Also note that there no assumptions can be made on the order of items
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/main/java/com/yahoo/document/DocumentUpdate.java]: Correct

File: document/src/main/java/com/yahoo/document/DocumentUpdate.java [hunk 2]
```diff
@@ -343,9 +374,28 @@ public void addAll(DocumentUpdate update) {
...
+
+    public FieldUpdate removeFieldUpdate(Field field) {
+        for (Iterator<FieldUpdate> it = fieldUpdates.iterator(); it.hasNext();) {
+            FieldUpdate fieldUpdate = it.next();
+            if (fieldUpdate.getField().equals(field)) {
+                it.remove();
+                return fieldUpdate;
+            }
+        }
+        return null;    }
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/main/java/com/yahoo/document/DocumentUpdate.java]: Correct

File: document/src/main/java/com/yahoo/document/DocumentUpdate.java [hunk 3]
```diff
@@ -343,9 +374,28 @@ public void addAll(DocumentUpdate update) {
...
      * @return the FieldUpdate previously at the specified position
      * @throws IndexOutOfBoundsException if index is out of range
+     * @deprecated use removeFieldUpdate(Field field) instead.
      */
+    @Deprecated
     public FieldUpdate removeFieldUpdate(int index) {
-        return fieldUpdates.remove(index);
+        FieldUpdate prev = getFieldUpdate(index);
+        fieldUpdates.remove(index);
+        return removeFieldUpdate(prev.getField());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/main/java/com/yahoo/document/DocumentUpdate.java]: Ooops, correct. Missing test...

File: document/src/main/java/com/yahoo/document/DocumentUpdate.java [hunk 4]
```diff
@@ -398,30 +448,27 @@ public String toString() {
         string.append(docId);
         string.append("': ");
         string.append("create-if-non-existent=");
-        string.append(createIfNonExistent.orElse(false));
+        string.append(createIfNonExistent);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/main/java/com/yahoo/document/DocumentUpdate.java]: Correct

File: document/src/main/java/com/yahoo/document/serialization/VespaDocumentDeserializer42.java [hunk 5]
```diff
@@ -472,7 +472,7 @@ public void read(DocumentUpdate update) {
 
         for (int i = 0; i < size; i++) {
             // TODO: Should use checked method, but doesn't work according to test now.
-            update.addFieldUpdateNoCheck(new FieldUpdate(this, update.getDocumentType(), serializationVersion));
+            update.addFieldUpdate(new FieldUpdate(this, update.getDocumentType(), serializationVersion));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/main/java/com/yahoo/document/serialization/VespaDocumentDeserializer42.java]: Correct

File: document/src/main/java/com/yahoo/document/serialization/VespaDocumentDeserializerHead.java [hunk 6]
```diff
@@ -29,7 +29,7 @@ public void read(DocumentUpdate update) {
 
         for (int i = 0; i < size; i++) {
             // TODO: Should use checked method, but doesn't work according to test now.
-            update.addFieldUpdateNoCheck(new FieldUpdate(this, update.getDocumentType(), 8));
+            update.addFieldUpdate(new FieldUpdate(this, update.getDocumentType(), 8));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply document/src/main/java/com/yahoo/document/serialization/VespaDocumentDeserializerHead.java]: Correct

**Pull request merged**